##author stephane.manciot@ebiznext.com
// license-header java merge-point
//
// Attention: Generated code! Do not modify by hand!
// Generated by: DefaultServiceException.vsl in andromda-spring-cartridge.
//
#set ($generatedFile = "${stringUtils.replace($service.fullyQualifiedDefaultExceptionName,'.','/')}.groovy")
#if ($stringUtils.isNotBlank($service.packageName))
package $service.packageName;
#end

import org.apache.commons.beanutils.PropertyUtils;
import org.springframework.validation.Errors

/**
 * The default exception thrown for unexpected errors occurring
 * within {@link $service.fullyQualifiedName}.
 */
class $service.defaultExceptionName
    extends java.lang.RuntimeException
{
    /** 
     * The serial version UID of this class. Needed for serialization. 
     */
    private static final long serialVersionUID = ${service.serialVersionUID}L;

    /**
     * Validation errors to be propagated to other layers
     */
    private Errors errors

    /**
     * The default constructor for <code>$service.defaultExceptionName</code>.
     */
    public ${service.defaultExceptionName}()
    {}

    /**
     * Constructs a new instance of <code>$service.defaultExceptionName</code>.
     *
     * @param throwable the parent Throwable
     */
    public ${service.defaultExceptionName}(Throwable throwable)
    {
        super(findRootCause(throwable));
    }

    /**
     * Constructs a new instance of <code>$service.defaultExceptionName</code>.
     *
     * @param message the throwable message.
     */
    public ${service.defaultExceptionName}(String message)
    {
        super(message);
    }

    /**
     * Constructs a new instance of <code>$service.defaultExceptionName</code>.
     *
     * @param errors the validation errors.
     */
    public ${service.defaultExceptionName}(Errors errors)
    {
        this.errors = errors
    }

    /**
     * Constructs a new instance of <code>$service.defaultExceptionName</code>.
     *
     * @param message the throwable message.
     * @param throwable the parent of this Throwable.
     */
    public ${service.defaultExceptionName}(String message, Throwable throwable)
    {
        super(message, findRootCause(throwable));
    }

    /**
     * Constructs a new instance of <code>$service.defaultExceptionName</code>.
     *
     * @param message the throwable message.
     * @param errors the validation errors
     * @param throwable the parent of this Throwable.
     */
    public ${service.defaultExceptionName}(String message, Errors, errors, Throwable throwable)
    {
        super(message, findRootCause(throwable))
        this.errors = errors
    }

    /**
     * @return validation errors if any, null otherwise
     */
    public Errors getErrors(){
        return this.errors
    }

    /**
     * Finds the root cause of the parent exception
     * by traveling up the exception tree
     */
    private static Throwable findRootCause(Throwable th)
    {
        if (th != null)
        {
            // Reflectively get any exception causes.
            try
            {
                Throwable targetException = null;

                // java.lang.reflect.InvocationTargetException
                String exceptionProperty = "targetException";
                if (PropertyUtils.isReadable(th, exceptionProperty))
                {
                    targetException = (Throwable)PropertyUtils.getProperty(th, exceptionProperty);
                }
                else
                {
                    exceptionProperty = "causedByException";
                    //javax.ejb.EJBException
                    if (PropertyUtils.isReadable(th, exceptionProperty))
                    {
                        targetException = (Throwable)PropertyUtils.getProperty(th, exceptionProperty);
                    }
                }
                if (targetException != null)
                {
                    th = targetException;
                }
            }
            catch (Exception ex)
            {
                // just print the exception and continue
                ex.printStackTrace();
            }

            if (th.getCause() != null)
            {
                th = th.getCause();
                th = findRootCause(th);
            }
        }
        return th;
    }
}